Costume index state

fsih opened the pull request:
Moves the selected costume index from internal state in costume-tab to redux state. This is in preparation for fixing https://github.com/LLK/scratch-gui/issues/1475

In order to have this work correctly, I needed to make updateEditingTarget, updateTargets, and setEditingCostumeIndex happen together atomically. This way, if a costume is deleted or the target sprite changes, the editing costume index and costume deletion happen together, instead of e.g. the costume index moving up first, then the data about the costume being deleted catching up later, which would confuse the paint editor.

I used thunk so that I could access the current state when determining what the action should do. (This is because our target state is sent all at once, instead of individual actions for each change, which means that in order to tell if a deletion has occurred, I need to compare the state against the previous state)

rschamp commented on src/reducers/targets.js:
```diff
@@ -21,34 +20,65 @@ const reducer = function (state, action) {
+const updateTargets = function (editingTarget, targetList) {
+    return function (dispatch, getState) {
+        const state = getState();
+        let editingCostume = state.editingCostumeIndex ? state.editingCostumeIndex : 0;
```
Hmm, now I see you probably named it that to be consistent with `editingTarget` which also makes sense... up to you if you want to change that or not.

rschamp commented on src/reducers/editing-costume-index.js:
```diff
@@ -0,0 +1,32 @@
+    case SET_EDITING_COSTUME_INDEX:
+        return action.index;
+    case UPDATE_TARGET_LIST:
+        return action.editingCostume;
```
Handling this external action here is a little unusual and potentially confusing. A comment here would be helpful.

fsih commented:
I added more comments in response to Ray's review. I also incorporated the suggestion from https://github.com/LLK/scratch-gui/pull/1582 (which was an alternative solution to this solution), which removed the need for all those promises! The change was, updateTargets updates costume index not only on delete costume, but on add costume, which means the costume tab no longer needs to listen on promises whenever it adds costumes in any way.

fsih commented:
@rschamp I was particularly wondering what your thoughts were on my use of thunk, and if it can/should be avoided. I didn't use thunk to make chains of actions like I think it's meant for, only to get access to the state in the reducer.

fsih commented:
I found that this change has some weird behavior ):
when the costume is changing rapidly in the project (forever next costume), and you add a costume, sometimes the paint editor doesn't switch to focusing on the new costume, but whichever random costume it happens to have cycled to.

fsih commented:
I found a crash that might be caused by the fact that we're updating editing target and targets separately, which I think would be fixed by this PR

https://github.com/LLK/scratch-gui/issues/1632

fsih commented:
Sorry about all the rambling on this PR, happy to talk in person...

fsih commented:
Going with the other solution for now
